File: Easy/LC872- Leaf-Similar-Trees/code.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:

        # function start
        def inorder(root, bst):
            if not root:
                return
            if not root.left and not root.right:
                bst.append(root.val)
            dfs(root.left, bst)
            dfs(root.right, bst)
        # function end

        bst1, bst2 = [], []
        dfs(root1, bst1)
        dfs(root2, bst2)

        if len(bst1) != len(bst2):
            return False
        for i in range(len(bst1)):
            if bst1[i] != bst2[i]:
                return False
        return True
        

File: Easy/LC872- Leaf-Similar-Trees/code.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void inorder(TreeNode* root, vector<int> &bst) {
        if(root == NULL) return;
        if(!root->left && !root->right)
            bst.push_back(root->val);
        inorder(root->left, bst);
        inorder(root->right, bst);
    }
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> bst1, bst2;
        inorder(root1, bst1);
        inorder(root2, bst2);

        if(bst1.size() != bst2.size()) {
            return false;
        }
        int i=0, n = bst1.size();
        while(i<n) {
            if(bst1[i] != bst2[i]) {
                return false;
            }
            i++;
        }
        return true;
    }
};
